### PR TITLE
DTP-711: Handle SequenceId edge cases

### PR DESCRIPTION
**Pull request by kaschula**
This PRs adds tests and validation handling for the sequenceId return from the configured sync function.

Conditions:
- [x] 1. SequenceId is `undefined`
- [x] 2. SequenceId is `'0'` and there is history in the channel. In which case history should be applied.

# comments

In scenario 1 where the sequenceId is `undefined`, the solution in this PR is throw an error and set the model state to `errored`. Additionally the `retryable()` logic has been updated to not run the callback function if the model is in the `errored`. Additionally if we catch an error while in the errored state then we re throw it, propagating it up the call stack. This is how the new `sequenceId is undefined error` will be surfaced to the user.

### CODE REVIEW

File: src/Model.test.ts [hunk 1]
```diff
@@ -1550,7 +1550,6 @@ describe('Model', () => {
     streams,
   }) => {
     const s1 = streams.newStream({ channelName: channelName });
-    s1.subscribe = vi.fn();
```
Comment (kaschula) [inline src/Model.test.ts:1553]: removed as this reassigned at line 1556

File: src/Model.test.ts [hunk 2]
```diff
@@ -1588,7 +1646,7 @@ describe('Model', () => {
         eventBufferOptions: defaultEventBufferOptions,
       },
     );
-    model.sync();
+    await model.sync();
```
Comment (kaschula) [inline src/Model.test.ts:1649]: @zknill I saw your name on this test. After I made my changes I was getting a random failure from this test I think its cause we call this sync function but don't await. Can you let me know if missing the await was on purpose. Thanks

File: src/Model.ts [hunk 3]
```diff
@@ -458,7 +470,14 @@ export default class Model<S extends SyncFuncConstraint> extends EventEmitter<Re
         lastSyncParams: this.lastSyncParams,
       });
       this.removeStream();
+
       const { data, sequenceId } = await this.syncFunc(...(this.lastSyncParams || ([] as unknown as Parameters<S>)));
+      if (sequenceId === undefined) {
+        const err = Error('sync function response: sequenceId is undefined');
```
Comment (zknill) [inline src/Model.ts:476]: nit-pick: I found this wording a little terse / hard to immediately understand. 
Do you think we could make it something that reads a bit more like a sentence and less like an error message 😅 maybe: 
> The sync function returned an undefined sequenceId

File: src/Model.test.ts (hunk 2 above)
Comment (zknill) [reply src/Model.test.ts:1649]: I probably just missed it when I wrote the test 👌 I don't see an obvious reason for it to be missing.

File: src/Model.ts (hunk 3 above)
Comment (kaschula) [reply src/Model.ts:476]: Good idea

File: src/stream/Middleware.ts [hunk 4]
```diff
@@ -169,6 +180,11 @@ export class OrderedHistoryResumer extends MiddlewareBase {
     // This is sufficiently low likelihood that this can be ignored for now.
     this.historicalMessages.sort((a, b) => this.reverseOrderer(a.id, b.id));
 
+    if (this.sequenceId === '0') {
+      // there will not be a message with ID = 0 so we don't bother searching through the history
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (zknill) [reply src/stream/Middleware.ts:186]: If it is true, then I think with seq = 0, we always hit a branch in this method where we exit immediately. So I wonder if we should do that check aboce `addHistoricalMessages(...)`. 

This suggests that that: 
- In all cases where seq=0, there will never be useful history. If so, there's no point even looking at the history?

[the remark this answers is not in the document]
Comment (kaschula) [reply src/stream/Middleware.ts:186]: So I guess its not strictly true just unlikely given that ADBC will produce messages that start with an ID of 1 due to the serial type on the 'id' column of the outbox table. I was trying to save us doing searches because after addHistoricalMessages is called we check the value of the sequenceId again and if its '0' we apply what ever history we have. 

We do this because '0' well always be lowest ID even if it exists the history it will still be applied. Thinking about it now I guess this is also an assumption so I could remove this check for now.

For context the solution to handling a sequenceId of zero is that we reply what ever history we have as seeking to a message with '0' id will likely never find anything and we assume a state with '0' is one that has never had any messages applied.

This is outline in the ticket comments: https://ably.atlassian.net/browse/DTP-711

File: src/stream/Middleware.ts [hunk 5]
```diff
@@ -145,10 +154,12 @@ export class OrderedHistoryResumer extends MiddlewareBase {
...
     if (messages.length === 0) {
       // If there were some messages in history then there have definitely been changes to the state
-      // and we can't reach back far enough to resume from the correct point.
+      // and we can't reach back far enough to resume from the correct point. If the sequenceId is
+      // '0' then we assume this due to SQL coalesc and no actual message will ever be found so
+      // flush() will reply what history is there/
       const noHistory = this.historicalMessages.length === 0;
       this.flush();
-      if (noHistory) {
+      if (noHistory || this.sequenceId == '0') {
```
Comment (mschristensen) [inline src/stream/Middleware.ts:162]:
```suggestion
      if (noHistory || this.sequenceId === '0') {
```

File: src/stream/Middleware.ts [hunk 6]
```diff
@@ -145,10 +154,12 @@ export class OrderedHistoryResumer extends MiddlewareBase {
     // the messages expired before the next page was requested.
     if (messages.length === 0) {
       // If there were some messages in history then there have definitely been changes to the state
-      // and we can't reach back far enough to resume from the correct point.
+      // and we can't reach back far enough to resume from the correct point. If the sequenceId is
+      // '0' then we assume this due to SQL coalesc and no actual message will ever be found so
+      // flush() will reply what history is there/
```
Comment (mschristensen) [inline src/stream/Middleware.ts:159]:
```suggestion
      // '0' then we assume this due to SQL coalesce and no actual message will ever be found so
      // flush() will reply what history is there.
```

File: src/stream/Middleware.ts [hunk 7]
```diff
@@ -169,6 +180,11 @@ export class OrderedHistoryResumer extends MiddlewareBase {
     // This is sufficiently low likelihood that this can be ignored for now.
     this.historicalMessages.sort((a, b) => this.reverseOrderer(a.id, b.id));
 
+    // A sequenceId of 0 means there is no historical message to seek to. All history should be applied after paging complete
```
Comment (mschristensen) [inline src/stream/Middleware.ts:183]: The logic is somewhat complex, so worth adding a clarifying comment.

When we have rewind-by-application-id this will be much simpler.

```suggestion
    // A sequenceId of 0 is a cursor that represents the position before the start of the stream.
    // There is no such historical message to seek to, instead we should paginate through all
    // of history to obtain all messages that were published since the sequenceId was obtained.
```

File: src/stream/Middleware.ts [hunk 8]
```diff
@@ -117,6 +117,15 @@ export class OrderedHistoryResumer extends MiddlewareBase {
     this.slidingWindow.subscribe(this.onMessage.bind(this));
   }
 
+  applyHistory() {
```
Comment (mschristensen) [inline src/stream/Middleware.ts:120]: We should mark this method `public` and move it alongside the other public methods

**Review by mschristensen — CHANGES_REQUESTED** [review]
Nice first PR! And great test coverage. Couple of comments

File: src/Model.ts [hunk 9]
```diff
@@ -469,7 +488,10 @@ export default class Model<S extends SyncFuncConstraint> extends EventEmitter<Re
       await this.retryable(this.syncRetryStrategy, fn);
     } catch (err) {
       this.logger.error('retries exhausted', { ...this.baseLogContext, action: 'resync()', err });
-      this.setState('errored', toError(err));
+      if (this.state !== 'errored') {
```
[the remark this answers is not in the document]
Comment (kaschula) [reply src/Model.ts:491]: Good Idea, I put this here cause I ran into exactly that situation. I will remove this one and move it to the `setState()`.
